### PR TITLE
Merging of additional combined params and eyelid

### PR DESCRIPTION
**Pull request by regzo2**
Here is the changelog for the additional Lip and Eye params, including a description of what each param does.

Here is also a Unity asset that includes a sample blinking controller and attached animations that will need to be edited or replaced. Adding it to your own animation controller should be as simple as adding the params to your controller and copying each layer from the included controller to your own controller setup. 

[BlinkingSetup.zip](https://github.com/benaclejames/VRCFaceTracking/files/7532334/BlinkingSetup.zip)

Added over 35 combined Lip parameters:

| Parameter | Description | Range |
| ------------- | ------------- | ------------- |
|JawOpenApe| Jaw opens with positive values, Mouth ape shapes with negative values| -1.0 - 1.0|
|JawOpenPuffRight| Jaw opens with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|JawOpenPuffLeft| Jaw opens with positive values, Left cheek puffs with negative values| -1.0 - 1.0|
|JawOpenSuck| Jaw opens with positive values, Cheeks suck in with negative values| -1.0 - 1.0|
|JawOpenForward| Jaw opens with positive values, Jaw moves forward with negative values| -1.0 - 1.0|
|||
|MouthUpperUpRightUpperInside| Upper right lip deflects to reveal teeth with positive values, Upper lips tuck into mouth with negative values| -1.0 - 1.0|
|MouthUpperUpRightPuffRight| Upper right lip deflects to reveal teeth with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|MouthUpperUpRightApe| Upper right lip deflects to reveal teeth with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|MouthUpperUpRightPout| Upper right lip deflects to reveal teeth with positive values, Mouth pouts with negative values| -1.0 - 1.0|
|MouthUpperUpRightOverlay| Upper right lip deflects to reveal teeth with positive values, Lower lip does an overlay shape with negative values| -1.0 - 1.0|
|||
|MouthUpperUpLeftUpperInside| Upper left lip deflects to reveal teeth with positive values, Upper lips tuck into mouth with negative values| -1.0 - 1.0|
|MouthUpperUpLeftPuffLeft| Upper left lip deflects to reveal teeth with positive values, Left cheek puffs with negative values| -1.0 - 1.0|
|MouthUpperUpLeftApe| Upper left lip deflects to reveal teeth with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|MouthUpperUpLeftPout| Upper left lip deflects to reveal teeth with positive values, Mouth pouts with negative values| -1.0 - 1.0|
|MouthUpperUpLeftOverlay| Upper left lip deflects to reveal teethwith positive values, Lower lip does an overlay shape  with negative values| -1.0 - 1.0|
|||
|MouthLowerDownRightLowerInside| Lower right lip deflects to reveal teeth with positive values, Lower lips tuck into mouth with negative values| -1.0 - 1.0|
|MouthLowerDownRightPuffRight| Lower right lip deflects to reveal teeth on the right with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|MouthLowerDownRightApe| Lower right  lip deflects to reveal teeth on the right with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|MouthLowerDownRightPout| Lower right lip deflects to reveal teeth on the right with positive values, Mouth pouts with negative values| -1.0 - 1.0|
|MouthLowerDownRightOverlay| Lower right lip deflects to reveal teeth with positive values, Lower lip does an overlay shape with negative values| -1.0 - 1.0|
|||
|MouthLowerDownLeftLowerInside| Lower left lip deflects to reveal teeth with positive values, Lower lips tuck into mouth with negative values| -1.0 - 1.0|
|MouthLowerDownLeftPuffLeft| Lower left lip deflects to reveal teeth with positive values, Left cheek puffs with negative values| -1.0 - 1.0|
|MouthLowerDownLeftApe| Lower left lip deflects to reveal teeth with positive values, Right cheek puffs with negative values| -1.0 - 1.0|
|MouthLowerDownLeftPout| Lower left lip deflects to reveal teeth with positive values, Mouth pouts with negative values| -1.0 - 1.0|
|MouthLowerDownLeftOverlay| Lower left lip deflects to reveal teeth with positive values, Lower lip does an overlay shape with negative values| -1.0 - 1.0|
|||
|SmileRightUpperOverturn| Right side of the mouth smiles with positive values, Upper lip does an overturn shape with negative values| -1.0 - 1.0|
|SmileRightLowerOverturn| Right side of the mouth smiles with positive values, Lower lip does an overturn shape with negative values| -1.0 - 1.0|
|SmileRightApe| Right side of the mouth smiles with positive values, Mouth ape shapes with negative values| -1.0 - 1.0|
|SmileRightOverlay| Right side of the mouth smiles with positive values, Lower lip does an overlay shape with negative values| -1.0 - 1.0|
|SmileRightPout| Right side of the mouth smiles with positive values, Mouth Pouts with negative values| -1.0 - 1.0|
|||
|SmileLeftUpperOverturn| Left side of the mouth smiles with positive values, Upper lip does an overturn shape with negative values| -1.0 - 1.0|
|SmileLeftLowerOverturn| Left side of the mouth smiles with positive values, Lower lip does an overturn shape with negative values| -1.0 - 1.0|
|SmileLeftApe| Left  side of the mouth smiles with positive values, Mouth ape shapes with negative values| -1.0 - 1.0|
|SmileLeftOverlay| Left side of the mouth smiles with positive values, Lower lip does an overlay shape with negative values| -1.0 - 1.0|
|SmileLeftPout| Left side of the mouth smiles with positive values, Mouth Pouts with negative values| -1.0 - 1.0|

Added several new bool params that act as a binary counting system (for an animation controller) for the Eyelids (possibly for now). These in particular work in a 4 bit space (or 16 steps), with an additional 2 bytes dedicated if you want to add Squeeze and Widened:
```
--Basic Blink--
LeftEyeLid1
LeftEyeLid2
LeftEyeLid4
LeftEyeLid8

RightEyeLid1
RightEyeLid2
RightEyeLid4
RightEyeLid8
```
```
--Blink & Eyes Widen--
**This is the setup that the included Unity asset uses
LeftEyeLidExpanded1
LeftEyeLidExpanded2
LeftEyeLidExpanded4
LeftEyeLidExpanded8
LeftEyeWidenToggle

RightEyeLidExpanded1
RightEyeLidExpanded2
RightEyeLidExpanded4
RightEyeLidExpanded8
RightEyeWidenToggle
```
```
--Blink, Squeeze, & Eyes Widen--
LeftEyeLidExpandedSqueeze1
LeftEyeLidExpandedSqueeze2
LeftEyeLidExpandedSqueeze4
LeftEyeLidExpandedSqueeze8
LeftEyeWidenToggle
LeftEyeSqueezeToggle

RightEyeLidExpandedSqueeze1
RightEyeLidExpandedSqueeze2
RightEyeLidExpandedSqueeze4
RightEyeLidExpandedSqueeze8
RightEyeWidenToggle
RightEyeSqueezeToggle
```

### CODE REVIEW

File: VRCFaceTracking/Params/Eye/EyeTrackingParams.cs [hunk 1]
```diff
@@ -22,7 +22,50 @@ public static class EyeTrackingParams
             new BoolEyeParameter(v2 => v2.Left.Openness < 0.5f, "LeftEyeLid"),
             new BoolEyeParameter(v2 => v2.Right.Openness < 0.5f, "RightEyeLid"),
             new BoolEyeParameter(v2 => (v2.Left.Openness + v2.Right.Openness)/2 < 0.5f, "CombinedEyeLid"),
-            
+
+            // Testing out the bool idea: Ideally would handle this in a class, but a test is a test.
+            new BoolEyeParameter(v2 => ((int)(((v2.Left.Openness * 15 + .5))%2) == 1), "LeftEyeLid1"),
```
Comment (benaclejames) [inline VRCFaceTracking/Params/Eye/EyeTrackingParams.cs:27]: Could this please be added to a new class for quarter-params? Thanks

**Review by benaclejames — CHANGES_REQUESTED** [review]
Please create classes for params with multiple states

File: VRCFaceTracking/Params/Eye/EyeTrackingParams.cs (hunk 1 above)
Comment (regzo2) [reply VRCFaceTracking/Params/Eye/EyeTrackingParams.cs:27]: I will see about that! Are the lip params good to go also?

Comment (benaclejames) [reply VRCFaceTracking/Params/Eye/EyeTrackingParams.cs:27]: Lip params look good for now! Perhaps I'll refactor them in the future, but for now they look fine!

**Review by regzo2 — COMMENTED** [review]
I've created BinaryEyeParameter to handle the binary parameters for eye data, and tidied up the parameters to use the new class. I've also added a new class to the ParamLib to handle BinaryEyeParameter and any future binary-type parameters (I will also do a pull request on ParamLib).

I've tested the new parameters on my own build and they seem to work as intended.

Let me know what you think! I'm open to adjusting the logic or making it more tidy if need be.

**Comment by benaclejames** [comment]
Right, but I physically can't compile this because BinaryParam isn't defined. Did you rename something or forget to commit something?

**Comment by regzo2** [comment]
You will need to update BaseParam.cs (ParamLib) as i've added BinaryParam to that, I opened a pull request for it if you would like to see that.

**Comment by regzo2** [comment]
https://github.com/benaclejames/ParamLib/pull/3/commits/62eae09a9ace79c502462f2efdebb989845871fc

**Comment by benaclejames** [comment]
I think the multiple bool param would be better in VRCFaceTracking, as opposed to in ParamLib. Since it's unlikely someone would need a similar thing for any other mod that modifies parameters. If you could include that commit in this PR, that'd be great!

**Comment by regzo2** [comment]
All righty sounds like a done deal!

**Comment by regzo2** [comment]
All right I'm not entirely sure how to commit the forked ParamLib submodule with my changes, i'm still trying to figure that out. I have never dealt with submodules on a repo before!

**Comment by benaclejames** [comment]
Don't commit a forked paramlib, since it's a submodule, it'd be in the paramlib repo. I'm saying you should put it in a file in VRCFaceTracking, It would probably fit well inside EyeParameter.cs

**Comment by regzo2** [comment]
Oh okay that makes it easy then.

**Comment by regzo2** [comment]
I will just put it in the ParamLib namespace in EyeParameter for now, may want to put it into a separate file eventually if I ever do this for lip params

The only other change I may need is making the BoolBaseParam constructor public since BinaryEyeParameter uses it to fulfill the BinaryParam constructor (pretty similar to how you have the XYEyeParameter create FloatBaseParams for its setup), unless I could work around that. I'll do another pull for just that change.

**Comment by regzo2** [comment]
Changing
`protected BoolBaseParam(string paramName) : base(paramName, VRCExpressionParameters.ValueType.Bool)
        {
        }`
to
`public BoolBaseParam(string paramName) : base(paramName, VRCExpressionParameters.ValueType.Bool)
        {
        }`
should allow you to build

**Comment by regzo2** [comment]
inside of BaseParam.cs

**Pull request merged**